### ISSUE TITLE
1.1.286: Inappropriate binding / comparison of parameters of inline expression

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#4395]
Starting in pyright 1.1.286 a false positive occurs in the following code; the expression for `s_either_inline2` is marked as an error:

```python
from typing import TypeVar, Generic, Callable, Union

T1 = TypeVar("T1")
T2 = TypeVar("T2")


class S(Generic[T1]):
    def __or__(self, other: "S[T2]") -> "S[Union[T1, T2]]": ...

def to(x: Callable[..., T1]) -> "S[T1]": ...
def or_(a: S[T1], b: S[T2]) -> S[Union[T1, T2]]: ...

s_int = to(int)
s_float = to(float)
s_either = s_int | s_float
s_either_inline1 = or_(to(int), to(float))


reveal_type(s_either)
reveal_type(s_either_inline1)

# in `to(x)`, x is bound to (...) -> int, which is seen 
# as incompatible with type[float]
s_either_inline2 = to(int) | to(float)  # ERROR

reveal_type(s_either_inline2)
```

The following output is produced:

```
No configuration file found.
pyproject.toml file found at /home/ry0/public-responsible-ai-toolbox.
Loading pyproject.toml file at /home/ry0/public-responsible-ai-toolbox/pyproject.toml
Assuming Python platform Linux
stubPath /home/ry0/public-responsible-ai-toolbox/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.286
/home/ry0/public-responsible-ai-toolbox/scratch/scratch.py
  /home/ry0/public-responsible-ai-toolbox/scratch/scratch.py:19:13 - information: Type of "s_either" is "S[int | float]"
  /home/ry0/public-responsible-ai-toolbox/scratch/scratch.py:20:13 - information: Type of "s_either_inline1" is "S[int | float]"
  /home/ry0/public-responsible-ai-toolbox/scratch/scratch.py:24:33 - error: Argument of type "Type[float]" cannot be assigned to parameter "x" of type "(...) -> T1@to" in function "to"
    Type "Type[float]" cannot be assigned to type "(...) -> int"
      Function return type "float" is incompatible with type "int"
        "float" is incompatible with "int" (reportGeneralTypeIssues)
  /home/ry0/public-responsible-ai-toolbox/scratch/scratch.py:26:13 - information: Type of "s_either_inline2" is "S[int]"
1 error, 0 warnings, 3 informations 
```

This is impacting some common code patterns involving the `SearchStrategy` class in the [hypothesis testing library](https://hypothesis.readthedocs.io/en/latest/).

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.
